ethcore/src/client/client.rs

```diff
@@ -1450,26 +1454,29 @@ impl ImportBlock for Client {
 		}
 
 		let raw = if self.importer.block_queue.is_empty() {
-			Some((
-				unverified.bytes.clone(),
-				unverified.header.hash(),
-				*unverified.header.difficulty(),
-			))
-		} else { None };
+			Some((unverified.bytes.clone(), *unverified.header.difficulty()))
+		} else {
+			None
+		};
 
 		match self.importer.block_queue.import(unverified) {
 			Ok(hash) => {
-				if let Some((raw, hash, difficulty)) = raw {
-					self.notify(move |n| n.block_pre_import(&raw, &hash, &difficulty));
+				if let Some((bytes, difficulty)) = raw {
+					self.notify(move |n| n.block_pre_import(&bytes, &hash, &difficulty));
 				}
 				Ok(hash)
 			},
 			// we only care about block errors (not import errors)
-			Err((block, EthcoreError::Block(err))) => {
-				self.importer.bad_blocks.report(block.bytes, format!("{:?}", err));
-				return Err(EthcoreError::Block(err))
+			Err((EthcoreError::Block(err), input)) => {
+				self.importer.bad_blocks.report(
+					input
+						.expect("BlockQueue::import enforces that `EthcoreError::Block` always returns Some(input); qed")
```

Comment:
Why not match `Err((EthcoreError::Block(err), Some(input))) => {` instead? Seems it would be more future proof. You could just print an error in case we get `(EthcoreError, None)` in the next match, so that you can easily detect if the assumption get's broken.
Current proof seems pretty weak to me.

Reply:
see https://github.com/paritytech/parity-ethereum/pull/11400#discussion_r370314948

I agree it's not future proof, but the right fix is to implement #11403. Feel free to disagree though.

Reply:
sure, the idea hit me too

Reply:
Well, you know I'm usually against panicking, especially if the assumptions are a bit weak like here: we assume something about internals of generic type and hope it will never change.

I'm more willing to just print an error in case we get `None`, I don't see how it is better to crash the node than to just not report a bad block. It seems that the risk of introducing the error is not worth the gains you could get from failing fast. The node will be in a perfect state if we don't panic and will continue to operate, the only thing we will miss is a report about bad block.

I fully agree though that #11403 is the right solution, but for now I'd rather avoid panicking.

Reply:
right, I don't have a strong opinion on that, but my reasoning was that if the panic would happen, it would urge the developers to implement #11403, otherwise it would just sit the backlog for ages

But either way is fine.